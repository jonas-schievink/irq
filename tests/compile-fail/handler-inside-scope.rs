//! Test that no handlers can be registered that are defined *inside* the scope they would be
//! registered for.

use irq::{handler, scope, scoped_interrupts};
use mock_pac::interrupt;

scoped_interrupts! {
    enum Interrupt {
        INT0,
    }

    use #[interrupt];
}

fn main() {
    let mut i = 0;

    scope(|scope| {
        handler!(int0 = move || i += 1);
        scope.register(Interrupt::INT0, int0);

        loop {
            // Idle loop
            break;
        }
    });
}
