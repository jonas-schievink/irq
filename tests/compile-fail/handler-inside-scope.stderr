error[E0716]: temporary value dropped while borrowed
  --> $DIR/handler-inside-scope.rs:19:9
   |
18 |     scope(|scope| {
   |            ----- has type `&irq::Scope<'1, Interrupt>`
19 |         handler!(int0 = move || i += 1);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ creates a temporary which is freed while still in use
20 |         scope.register(Interrupt::INT0, int0);
   |         ------------------------------------- argument requires that borrow lasts for `'1`
...
26 |     });
   |     - temporary value is freed at the end of this statement
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0597]: `closure` does not live long enough
  --> $DIR/handler-inside-scope.rs:19:9
   |
18 |     scope(|scope| {
   |            ----- has type `&irq::Scope<'1, Interrupt>`
19 |         handler!(int0 = move || i += 1);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         borrowed value does not live long enough
   |         argument requires that `closure` is borrowed for `'1`
...
26 |     });
   |     - `closure` dropped here while still borrowed
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
